CoffeeShop.playground/Contents.swift

```
@@ -315,17 +328,65 @@ class CoffeeShopAppTest: XCTestCase {
         XCTAssertFalse(isFilename)
     }
 
-    func test_test_isNotFilenameNoName() {
+    func test_isNotFilenameNoName() {
         let isFilename = CoffeeShopApp.isFilename(".csv")
         XCTAssertFalse(isFilename)
     }
 
-    func test_test_isNotFilenameNoExtension() {
+    func test_isNotFilenameNoExtension() {
         let isFilename = CoffeeShopApp.isFilename("coffee_shop.")
         XCTAssertFalse(isFilename)
     }
+
+    func test_3ClosestCoffeeShops() {
+        let app = CoffeeShopApp(["47.6", "-122.4", "coffee_shop.csv"])
+        XCTAssertNotNil(app)
+        let fileContent = """
+                        Starbucks Seattle,47.5809,-122.3160
+                        Starbucks SF,37.5209,-122.3340
+                        Starbucks Moscow,55.752047,37.595242
+                        Starbucks Seattle2,47.5869,-122.3368
+                        Starbucks Rio De Janeiro,-22.923489,-43.234418
+                        Starbucks Sydney,-33.871843,151.206767
+                        """
+
+        guard let coffeeShopsWithDistance = app!.closestCoffeeShops(count: 3, coffeeShopsCSV: fileContent) else {
+            NSLog("Invalid data format")
+            return
+        }
+
+        for coffeeShopWithDistance in coffeeShopsWithDistance {
+            let distance = String(format: "%.4f", coffeeShopWithDistance.distance)
+            print("\(coffeeShopWithDistance.coffeeShop.name),\(distance)")
+        }
+    }
+
 }
 
 CoffeeShopAppTest.defaultTestSuite.run()
 
+var commandLineArguments = CommandLine.arguments
+commandLineArguments.removeFirst()
+
+done: if let app = CoffeeShopApp(commandLineArguments) {
+
+    let fileUrl = playgroundSharedDataDirectory.appendingPathComponent(app.shopDataFilename)
+    guard let fileContent = try? String(contentsOf: fileUrl) else {
+        NSLog("Unable to read \(app.shopDataFilename) from ~\\Documents\\Shared Playground Data")
+        break done
```

I don't think this is required, I would expect to throw a custom exception if the input is invalid instead of using this goto like statement.

---

By "Malformed entries should cause the program to exit appropriately" request I thought that the app should not throw and error, that's why I logged the messages in the console. I'll refactor these parts

---

`exit appropriately` can mean that you throw a explicit error if the user input is not right.

The best explanation I've found about this pattern is in the rust documentation file:///home/calin/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/book/2018-edition/ch09-03-to-panic-or-not-to-panic.html, panic! is like throwing an exception.